hugr-core/src/types/type_row.rs

```diff
@@ -71,7 +71,7 @@ impl<RV: MaybeRV> TypeRowBase<RV> {
     /// Applies a substitution to the row.
     /// For `TypeRowRV`, note this may change the length of the row.
     /// For `TypeRow`, guaranteed not to change the length of the row.
-    pub(super) fn substitute(&self, s: &Substitution) -> Self {
+    pub(crate) fn substitute(&self, s: &Substitution) -> Self {
```

Comment:
Hmmm, all the substitute methods take an `&self`. I guess we were thinking that the main use case was producing a substitution of an OpDef's signature given type arguments, so it made sense to do the cloning at the same time. I think I could be persuaded to change to an `&mut self` but I'm not sure it's much advantage. Monomorphizing involves copying the polymorphic func to make a monomorphic instantiation, too.

Reply:
Yes I just realized that; I am cloning each op in `monomorphize.rs` before calling substitute.. I should at least be consistent... 🤦

Reply:
Ah, but that prevents the default impl of just returning self in the trait (unless we require `Clone`)

Reply:
So the difference here is that the thing we are copying here is likely the whole Hugr; if one copied the ops whilst applying the substitution, one would then still need to copy the edges. Hence, instead one first clones the whole Hugr, and then a mutating substitution on the ops becomes appropriate....

Reply:
However this is only useful given a way to get an `&mut OpType`. If we have only `&OpType` and `replace_op` then `substitute(&self,..) -> Self` is as good as anything. And we have `replace_op` because it allows us to check the replacement node is of the correct type (as per static guarantee *à la* `RootChecked`)

Reply:
Standardized all `substitute` methods to take `&self`. We could switch to a `subst_mut` in the future if we had more call to it - like a copying method for subgraphs that preserves incoming edges - but we don't ATM so I'm skipping for now. (I guess it might be a breaking change but only to the level of requiring clients calling the existing `substitute` to import a trait.)